### PR TITLE
mgr/dashboard_v2: Added tcmu iSCSI controller.

### PR DESCRIPTION
**Pull request by sebastian-philipp**
* Based on `mgr/dashboard/rbd_iscsi.py`.
* Fixed Python3 compatibility issues.
* Added test.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

### CODE REVIEW

**Review by p-se — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        mgr_mock.list_servers.return_value = mocked_servers
+        mgr_mock.get_metadata.return_value = mocked_metadata
+        mgr_mock.get_daemon_status.return_value = mocked_get_daemon_status
+        mgr_mock.get_counter.return_value = mocked_get_counter
+        mgr_mock.get_rate.return_value = mocked_get_rate
+        mgr_mock.url_prefix = ''
+        TcmuIscsi.mgr = mgr_mock
+        TcmuIscsi._cp_config['tools.authenticate.on'] = False  # pylint: disable=protected-access
+
+        cherrypy.tree.mount(TcmuIscsi(), "/api/test/rgw")
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: Consider using /api/test/iscsi here :-)

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tox.ini [hunk 2]
```diff
@@ -37,7 +37,7 @@ commands =
     bash ../src/vstart.sh -n
     sleep 10
     python ./bin/ceph config-key set mgr/dashboard_v2/x/server_port 9865
-    python ./bin/ceph mgr module enable dashboard_v2
+    python ./bin/ceph mgr module enable dashboard_v2 --force
```
Comment (p-se) [inline src/pybind/mgr/dashboard_v2/tox.ini]: I've been able to fix this issue by installing the dependencies of `requirements.txt` in the `dashboard_v2` folder using both, `pip2` and `pip3`.

Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tox.ini]: I guess you added this here to be able to load the module even without having the `python-coverage` package install. I don't think this is a good idea because this will also prevent us from catching errors from other reasons. It's very important that we can test if the module is correctly enabled.

File: src/pybind/mgr/dashboard_v2/tox.ini [hunk 3]
```diff
@@ -55,7 +55,7 @@ commands =
     sleep 5
     bash ../src/stop.sh
     rm {toxinidir}/ceph.conf
-    mv {toxinidir}/../../../../build/.coverage.mgr {toxinidir}/
+    test -f mv {toxinidir}/../../../../build/.coverage.mgr && mv {toxinidir}/../../../../build/.coverage.mgr {toxinidir}/
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tox.ini]: if the module can be enabled without the `--force` then you don't need this.

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tox.ini (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tox.ini]: I had a different error. Mine looked like a non-deterministic error to me.

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tox.ini]: No. This was a non-deterministic error. Often, disabling and enabling the dashboard_v2 did fix the issue for me. But I'll check.

Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/tox.ini]: Nonetheless, we should not pass the `--force` flag, it's preferable to have spurious false failures than not catching real failures

**Review by p-se — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/tcmu_iscsi.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+SERVICE_TYPE = 'tcmu-runner'
+
+
+@ApiController('tcmuiscsi')
+@AuthRequired()
+class TcmuIscsi(RESTController):
+    # pylint: disable=too-many-locals,too-many-nested-blocks
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def default(self, *args, **kwargs):  # pylint: disable=unused-argument
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/controllers/tcmu_iscsi.py]: If this class extends `RESTController`, why not rename this method to `list` and remove the decorators. Also, no need to declare the `*args` and `**kwargs` parameters if they are not used

File: src/pybind/mgr/dashboard_v2/tests/test_auth.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 from ..tools import Session
 
 
+@skipIf(bool(os.environ.get('SKIP_LOADED_BY_MGR', False)), 'Needs to run in mgr')
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/tests/test_auth.py]: This code is duplicated many times in this PR, so it seems to me a good candidate for a new decorator.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tests/test_auth.py (hunk 5 above)
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tests/test_auth.py]: I don't think we should add this. If some developer wants to run a specific test he can do it using py.test without tox.

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+mocked_get_rate = {
+    'librbd-42-pool1-image1.rd': 43,
+    'librbd-42-pool1-image1.wr': 44,
+    'librbd-42-pool1-image1.rd_bytes': 45,
+    'librbd-42-pool1-image1.wr_bytes': 46,
+}
+
+
+class TcmuIscsiControllerTest(ControllerTestCase, CPWebCase):
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: This test should not extend from `ControllerTestCase`, in this test you should access the controller through cherrypy.test functions like `getPage` and others.

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        mgr_mock.get_counter.return_value = mocked_get_counter
+        mgr_mock.get_rate.return_value = mocked_get_rate
+        mgr_mock.url_prefix = ''
+        TcmuIscsi.mgr = mgr_mock
+        TcmuIscsi._cp_config['tools.authenticate.on'] = False  # pylint: disable=protected-access
+
+        cherrypy.tree.mount(TcmuIscsi(), "/api/test/tcmu")
+
+    def __init__(self, *args, **kwargs):
+        super(TcmuIscsiControllerTest, self).__init__(*args, dashboard_port=54583, **kwargs)
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py:65]: If you access the controller using cherrypy helper functions, you don't need to pass the dashboard port.

File: src/pybind/mgr/dashboard_v2/tests/helper.py [hunk 8]
```diff
@@ -24,12 +24,16 @@ class ControllerTestCase(unittest.TestCase):
     DASHBOARD_PORT = os.environ.get('DASHBOARD_V2_PORT', 8080)
 
     def __init__(self, *args, **kwargs):
+        self.dashboard_host = kwargs.pop('dashboard_host') \
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tests/helper.py:27]: I don't think we need this for now. I understand you required this because of the iscsi test case, but as I commented previously you can avoid this.

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py (hunk 6 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: I'd like to keep other `vstart.sh-tests` and this test as similar as possible. Tbh, the functionality provided by `ControllerTestCase` is better than `CPWebCase` and I'm very much in favour of using all goodies of `ControllerTestCase`. 

Sure, I could resurrect parts of `ControllerTestCase` from the day before yesterday, but for which reason? This test could suddenly behave differently than other tests, and I don't like that.

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py (hunk 7 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py:65]: see above.

File: src/pybind/mgr/dashboard_v2/tests/helper.py (hunk 8 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tests/helper.py:27]: see above.

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py (hunk 6 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: ok

File: src/pybind/mgr/dashboard_v2/tests/test_auth.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tests/test_auth.py]: it doesn't hurt anyone, but it provides a way to execute some tests without thinking about an existing vstart.sh cluster.

**Review by votdev — DISMISSED** [review]

**Review by p-se — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/__init__.py [hunk 9]
```diff
@@ -25,3 +25,5 @@ def __getattr__(self, item):
     logging.basicConfig(level=logging.DEBUG)
     logger = logging.getLogger(__name__)
     logging.root.handlers[0].setLevel(logging.DEBUG)
+    os.environ['PATH'] = '{}:{}'.format(os.path.abspath('../../../../build/bin'),
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/__init__.py]: Could you add this change to a single commit?

**Review by rjfd — DISMISSED** [review]
lgtm

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+mocked_get_counter = {
+    'librbd-42-pool1-image1.lock_acquired_time': [[10000.0, 10000.0]],
+    'librbd-42-pool1-image1.rd': 43,
+    'librbd-42-pool1-image1.wr': 44,
+    'librbd-42-pool1-image1.rd_bytes': 45,
+    'librbd-42-pool1-image1.wr_bytes': 46,
+}
+
+mocked_get_rate = {
```
Comment (ricardoasmarques) [inline src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: Apparently this mock is incorrect.

Looking into ["get_rate" code](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/module.py#L247), I'd say that this method should return a number, and not a dictionaty. Am I wrong?

File: src/pybind/mgr/dashboard_v2/__init__.py (hunk 9 above)
Comment (jecluis) [reply src/pybind/mgr/dashboard_v2/__init__.py]: I agree. This seems a different logical change from the rest of the patch.

File: src/pybind/mgr/dashboard_v2/tests/helper.py [hunk 11]
```diff
@@ -100,3 +104,7 @@ def _rbd_cmd(cls, cmd):
         _cmd = ['rbd']
         _cmd.extend(cmd)
         return cls._cmd(_cmd)
+
+
+skip_if_loaded_by_mgr = unittest.skipIf(bool(os.environ.get('SKIP_LOADED_BY_MGR', False)),
```
Comment (jecluis) [inline src/pybind/mgr/dashboard_v2/tests/helper.py]: This seems to suggest that we will be skipping unit tests if we were "loaded by the mgr"? Why?

File: src/pybind/mgr/dashboard_v2/tests/test_auth.py [hunk 12]
```diff
@@ -6,12 +6,11 @@
...
 from unittest import skipIf
 
-from mock import patch
-
-from .helper import ControllerTestCase
+from .helper import ControllerTestCase, skip_if_loaded_by_mgr
 from ..tools import Session
 
 
+@skip_if_loaded_by_mgr
```
Comment (jecluis) [inline src/pybind/mgr/dashboard_v2/tests/test_auth.py]: Regardless of whether this whole "skip if loaded by the mgr" actually has some kernel of sense to it, adding these decorators is a different logical change from the rest of the patch, and should be a different commit.

File: src/pybind/mgr/dashboard_v2/tests/test_block_pool.py [hunk 13]
```diff
@@ -2,9 +2,10 @@
 
 from __future__ import absolute_import
 
-from .helper import ControllerTestCase, authenticate
+from .helper import ControllerTestCase, authenticate, skip_if_loaded_by_mgr
 
 
+@skip_if_loaded_by_mgr
```
Comment (jecluis) [inline src/pybind/mgr/dashboard_v2/tests/test_block_pool.py]: same for this file - and all the other files, really, that were modified solely to add this decorator and its import: move them all to a a different patch.

**Review by jecluis — DISMISSED** [review]
This patch is actually AT LEAST two (or maybe even three) logical patches, and needs to be split.

Additionally, the whole thing about skipping tests is introduced during a seemingly unrelated patch, and is not justified, at all. Some additional information in the code, may it be in the form of comments, for instance, or even in the commit message, would be appreciated.

**Comment by sebastian-philipp** [comment]
> Additionally, the whole thing about skipping tests is introduced during a seemingly unrelated patch, and is not justified, at all

I've split the commit. Fixing the tests is of course justified.

**Comment by jecluis** [comment]
@sebastian-philipp you misunderstand: the test skipping is, in fact, unjustified until now because you have not justified it anywhere, or provided any context as to why skipping them (or the decorator) is required, needed, whatever.
I'm not asserting it's not justified. I'm stating that you have not provided a reason for doing it, somewhere where people can actually check when/if this blows up in someone's face.

**Review by p-se — DISMISSED** [review]

**Comment by jecluis** [comment]
@sebastian-philipp do you mind opening a different pull request for the commits that are not part of this PR's purpose (i.e., not addressing the tcmu-runner iscsi stuff)?

**Comment by sebastian-philipp** [comment]
@jecluis PR merge commits will vanish if we rebase next time. The end result will be the same.

**Comment by jecluis** [comment]
@sebastian-philipp ?

**Comment by jecluis** [comment]
@sebastian-philipp let me rephrase, because I think you misunderstood the purpose of my last request:

Can you split this PR in two distinct PRs, given this PR is addressing various different things that are not (apparently) related?

By having different PRs we can have different discussions on different items, without holding the one or the other.

Also, regardless of how things eventually end up in the tree, it is good manner to keep PRs as logical units. Think of them like law Bills, right? Why would we include unrelated items in a given Bill, when the title on the Bill and the discussion being held is something else? (although, typically, when lawmakers pull those stunts, they do that in order to include legislation without anyone noticing or paying too much attention, or simply because not approving said legislation would likely mean not approving the larger, apparently more important one).

**Comment by sebastian-philipp** [comment]
> Can you split this PR in two distinct PRs, given this PR is addressing various different things that are not (apparently) related?

They are related together. The `skip_if_not_loaded_by_mgr` addition is especially useful for the iSCSI controller. Being able to execute the unit tests for this controller without a running vstart cluster was very helpful. I surely wouldn't have added these decorators, if they we not helpful for testing this PR. 

> Also, regardless of how things eventually end up in the tree, it is good manner to keep PRs as logical units. Think of them like law Bills, right? Why would we include unrelated items in a given Bill, when the title on the Bill and the discussion being held is something else? (although, typically, when lawmakers pull those stunts, they do that in order to include legislation without anyone noticing or paying too much attention, or simply because not approving said legislation would likely mean not approving the larger, apparently more important one).

Independent of the fact that these changes are in fact related. 

This PR is needed for a frontend PR that is mandatory for the initial PR. That initial PR is scheduled for the end of this week and I'm very much in favor of keeping this date. Having two unrelated commits in one PR may be unlawful for ceph/ceph, but it won't matter in any way for the upstream PR.

**Comment by rjfd** [comment]
@sebastian-philipp 
> Being able to execute the unit tests for this controller without a running vstart cluster was very helpful. I surely wouldn't have added these decorators, if they we not helpful for testing this PR.

As I said before (https://github.com/openattic/ceph/pull/82#discussion_r167208663), you can run the tests individually and thus your argument does not apply.

I agree with @jecluis, the `skip` decorator changes, and the `path` changes, should be moved to a different PR, as they are not even required for the initial PR!

**Comment by sebastian-philipp** [comment]
> I agree with @jecluis, the skip decorator changes, and the path changes, should be moved to a different PR, as they are not even required for the initial PR!

I'm getting tons of test failures when executing these tests;

```
PATH=...snip... PYTHONPATH=...snip...  UNITTEST=true LD_LIBRARY_PATH=...snip...  WEBTEST_INTERACTIVE=false py.test tests/
============================= test session starts ==============================
platform linux2 -- Python 2.7.14, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: ...snip... , inifile:
plugins: cov-2.5.1
collected 51 items

tests/test_auth.py FFFFFFFF                                              [ 15%]
tests/test_block_pool.py F                                               [ 17%]
tests/test_cluster.py ss                                                 [ 21%]
tests/test_dashboard.py FF                                               [ 25%]
tests/test_host.py F                                                     [ 27%]
tests/test_nodb.py ........................                              [ 74%]
tests/test_notification.py ...                                           [ 80%]
tests/test_rgw.py FF                                                     [ 84%]
tests/test_settings.py FF                                                [ 88%]
tests/test_tools.py ......                                               [100%]
... lots of test failures...
================== 16 failed, 33 passed, 2 skipped in 7.07 seconds ==================
```

I think providing a way to successfully run theses tests should be part of our initial PR.

**Comment by rjfd** [comment]
@sebastian-philipp 
> I think providing a way to successfully run theses tests should be part of our initial PR.

They run fine in our jenkins builds and in other peoples' environments. You should investigate why is not run successfully in your computer.

**Comment by jecluis** [comment]
@sebastian-philipp considering your last comment and @rjfd's comment, I am confused about something though: if the problem is tests failing on your machine, what is the benefit of skipping tests by annotating them?

**Comment by sebastian-philipp** [comment]
In `dashboard_v2` we have four different kinds of tests. All of them fit very well together and all of them are needed: 

## 1. Python Unit Tests

These tests are independent and free of all side effects. For example, we can test internal library functions by simply validating their documentation:
```python
def build_url(host, scheme=None, port=None):
    """
    Build a valid URL. IPv6 addresses specified in host will be enclosed in brackets
    automatically.

    >>> build_url('example.com', 'https', 443)
    'https://example.com:443'

    >>> build_url(host='example.com', port=443)
    '//example.com:443'

    >>> build_url('fce:9af7:a667:7286:4917:b8d3:34df:8373', port=80, scheme='http')
    'http://[fce:9af7:a667:7286:4917:b8d3:34df:8373]:80'
    """
    pass
```
If you modify the function and forget the colon before the port, this test will fail. `tests/test_tools.py RESTControllerTest` is another example of a test without any side effects or external dependencies. 

Running these tests is fast. It takes about one second to run them all. This is perfect for test driven development, as you can fix a bug an run all these tests basically instantaneous. 

You can use `mock` to develop these tests in order to make them independent and side effect free. 

But John was right. We should not mock the cluster, as changes in the cluster then get unnoticed. Therefore, we have a second layer of tests:

## 2. API Tests. 

We already had them in openATTIC standalone in a sub-directory `gatling` maintained by @s0nea . These tests are executed against a running Ceph cluster. They will have side effects (e.g. creating and deleting pools, rbds, crush rules, ...). Of course, executing them takes at one magnitude longer. About 6½ minutes on my system. Thus, you're running these tests only before committing your changes. We have plenty of those to test our controllers. 

Note that you need to reload the `dashboard_v2` mgr module every time you want to re-test your code.

Another downside is that you need to reconfigure `cmake` in order to change the Python version from 2.7 to 3. But having them is important and we depend on them.

## 3. Frontend unit tests

They have a similar role as our Python unit tests.

## 4. End-to-end tests

This is basically testing everything together. Frontend, API, Backend, and the cluster. Basically instructing the browser to click on a button and then wait, unit an (e.g.) RBD is created with the correct properties.

---

Python unit tests and API tests are not the same thing, even though they are located in a common `test/` directory and are executed in the same testing framework.

**Comment by rjfd** [comment]
@sebastian-philipp thanks for the detailed summary on different tests kinds. This really shows that including test changes in this PR, which is entitled "Added tcmu iSCSI controller.", does not make sense, since it's delaying the merge of the changes that match the PR's title.

Opening a new PR would be much better for having a discussion on how to configure our tests execution.

**Comment by sebastian-philipp** [comment]
I guess we have a loop now. 

As I said before, both changes are related together. The `skip_if_not_loaded_by_mgr` addition is especially useful for the iSCSI controller: Being able to execute the unit tests without the API tests. This PR only adds stuff that is needed for developing the iSCSI controller. 

And secondly:

This PR is needed for a frontend PR that is mandatory for the initial PR. That initial PR is scheduled for the end of this week and I'm very much in favor of keeping this date. Having two unrelated commits in one PR may be unlawful for ceph/ceph, but it won't matter in any way for the upstream PR.

**Review by ricardoasmarques — APPROVED** [review]
lgtm

**Review by jecluis — APPROVED** [review]
